crates/bevy_math/src/shape_sampling.rs

```diff
@@ -0,0 +1,239 @@
+use std::f32::consts::{PI, TAU};
+
+use crate::{primitives::*, Vec2, Vec3};
+use rand::Rng;
+
+/// Exposes methods to uniformly sample a variety of primitive shapes.
+pub trait ShapeSample {
+    /// The type of vector returned by the sample methods, [`Vec2`] for 2D shapes and [`Vec3`] for 3D shapes.
+    type Output;
+
+    /// Uniformly sample a point from inside the area/volume of this shape, centered on 0.
+    ///
+    /// # Example
+    /// ```
+    /// # use bevy_math::prelude::*;
+    /// let square = Rectangle::new(2.0, 2.0);
+    ///
+    /// // Returns a Vec2 with both x and y between -1 and 1.
+    /// println!("{:?}", square.sample_volume(&mut rand::thread_rng()));
+    /// ```
+    fn sample_volume(&self, rng: &mut impl Rng) -> Self::Output;
```

Comment:
I think you don't need to target `Rng` but can use `RngCore` instead, as `Rng` will apply to `RngCore` as an auto-trait. I'd also suggest using a generic `<R: RngCore + ?Sized>` instead as `RngCore` is implemented for both structs and `&mut` references, which should in turn make the interface nicer to use.

Reply:
I'm wondering if there is any advantage to taking `RngCore` rather than `Rng` because we need to depend on `rand` not `rand_core` (due to `gen_range` etc.) and anything implementing `RngCore` implements `Rng` (which we need because those are the methods we are using)?

Reply:
The `rand` [docs](https://docs.rs/rand/latest/rand/trait.Rng.html) suggest the current usage `fn foo<R: Rng + ?Sized>(rng: &mut R)`.